Add state dict for LSTM and RNNCell and helper functions for accessing weights and bias

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40333 Add state dict for LSTM and RNNCell and helper functions for accessing weights and bias**

Add state_dict support for dynamic quantized LSTM/GRU/RNNCell.

Add helper functions get_weight and get_bias for LSTM and RNNCells

Differential Revision: [D22151020](https://our.internmc.facebook.com/intern/diff/D22151020/)

raghuramank10000 committed:
Add state dict for LSTM and RNNCell and helper functions for accessing weights and bias

Add state_dict support for dynamic quantized LSTM/GRU/RNNCell.

Add helper functions get_weight and get_bias for LSTM and RNNCells

Differential Revision: [D22151020](https://our.internmc.facebook.com/intern/diff/D22151020/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Add state dict for LSTM and RNNCell and helper functions for accessing weights and bias"


Add state_dict support for dynamic quantized LSTM/GRU/RNNCell.

Add helper functions get_weight and get_bias for LSTM and RNNCells

Differential Revision: [D22151020](https://our.internmc.facebook.com/intern/diff/D22151020/)

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantized_module.py:
```diff
@@ -864,3 +950,5 @@ def test_cell_api(self):
                 self.assertEqual(result[0], result_module[0], msg="RNNCell module API failed")
                 self.assertEqual(result[1], result_module[1], msg="RNNCell module API failed")
+                check_weight_bias_api(cell_dq)
+                check_serialization(cell_dq, cell_dict[rnn_type](**kwargs))
```
Can we factor out `check_weight_bias_api` and `check_serialization` into common_quantization since it is being used in multiple places?

supriyar commented on torch/nn/quantized/dynamic/modules/rnn.py:
```diff
@@ -285,6 +285,32 @@ def quantize_and_pack(w, b):
+                suffix = '_reverse' if direction == 1 else ''
+                key_name1 = 'weight_ih_l{layer_idx}{suffix}'.format(layer_idx=layer, suffix=suffix)
+                key_name2 = 'weight_hh_l{layer_idx}{suffix}'.format(layer_idx=layer, suffix=suffix)
+                packed_weight_bias = self._all_weight_values[count].param.__getstate__()[0][4]
```
Can you add some comments here explaining the choice of the indexes? Is this accessing the tuple `CellParamsSerializationType`?

supriyar reviewed (approved):
Left a couple of comments, please address before landing. Thanks

raghuramank10000 committed:
Update on "Add state dict for LSTM and RNNCell and helper functions for accessing weights and bias"


Add state_dict support for dynamic quantized LSTM/GRU/RNNCell.

Add helper functions get_weight and get_bias for LSTM and RNNCells

Differential Revision: [D22151020](https://our.internmc.facebook.com/intern/diff/D22151020/)

[ghstack-poisoned]
